Consolidate type_utils in a way that can be reused in substitution results that need to be coerced to a specific type

ivanpauno opened the pull request:
Follow up of https://github.com/ros2/launch_ros/pull/137.

This try to consolidate the `launch.frontend.type_utils` module, in a way it's usable from custom actions/descriptions that requires an user to specify the type.

To consolidate the coercion rules allowed I've dropped support of `Union[...]` and non uniform lists, we really don't need that.

I have to consolidate documentation (partially outdated), type annotations and add more test cases.

This could also be used in the future in places like:
- https://github.com/ros2/launch/blob/24ae1327b4fd6c358c40899569ceabbdb4bc57f4/launch/launch/actions/timer_action.py#L59

so we always use the same rules to convert result of substitutions to specific types.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+            x if isinstance(x, str) else  # Don't convert strings to TextSubstitution
+            normalize_to_list_of_substitutions(cast(SomeSubstitutionsType, x)) for x in value
+        ])
+    if types_in_list.issubset({bool, Substitution}):
```
@ivanpauno if `types_in_list == {float, int, bool, Substitution}`, this will not fail until the "typed" substitution is performed. Is that intentional?

ivanpauno replied:
Why?
It should fail right away.

ivanpauno replied:
According to the tests I've added, that case is failing right away.

hidmic replied:
Uh, I'm curious where and how it fails.

hidmic commented on launch/test/launch/utilities/test_type_utils.py:
```diff
@@ -0,0 +1,149 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
```
@ivanpauno this is missing some tests e.g. for `normalize_typed_substitution()` and `perform_typed_substitution()`.

ivanpauno replied:
this is missing a lot of tests :joy:

ivanpauno replied:
Ok, I think those functions are well covered now.

hidmic commented on launch/launch/frontend/parser.py:
```diff
@@ -83,11 +85,44 @@ def parse_action(self, entity: Entity) -> (Action, Tuple[Any]):
+        self, value: StrSomeValueType
+    ) -> NormalizedValueType:
+        """
+        Parse substitutions in `value` if there are some.
```
@ivanpauno nit: 
```suggestion
        Parse substitutions in `value` if there are any.
```

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+
+def get_typed_value(
+    value: Union[Text, List[Text]],
+    data_type: Optional[AllowedTypesType],
```
no, `int` is also valid here for example.

ivanpauno replied:
take a look to the new test cases

hidmic replied:
Oops, yeah, I misread the code initially.

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+        f", expected a list of '{data_type}'. value='{value}'"
+    )
+    if types_in_list.issubset({str, Substitution}):
+        # list of mixed strings and substitutions
```
That case is checking for a single substitution, this one is a list of substitutions and strings.

PS: This comment makes me think that I'm not handling a list of only substitutions correctly... I will solve that after adding tests.

hidmic replied:
I may be crazy, but `is_substitution()` implementation above appears to cover substitution lists that can be normalized:
```python
def is_substitution(x):
    return (
        isinstance(x, Substitution) or
        (
            isinstance(x, collections.abc.Iterable) and
            len(x) > 0 and
            all(isinstance(item, (Substitution, str)) for item in x) and
            any(isinstance(item, Substitution) for item in x)
        )
    )
```

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+    List[int], List[float], List[bool], List[str]
+]]
+"""Allowed types."""
+AllowedTypesType = Union[ScalarTypesType, ListTypesType]
```
> I understand why the focus on uniform lists, but this is clearly at odds with Parser.get_attr() claiming it can handle non-uniform lists.

Non uniform lists add extra complexity, and I can't think of an use case for them.
I have to update old doc blocks.

ivanpauno replied:
> Hmm, a mixed integer/float

The coercion functionality is flexible, and a substitution without a trailing `.` can be coerced to a `float`:
https://github.com/ros2/launch/blob/7cc161161eaff98b1f6cd3068148af9ed29fea00/launch/test/launch/utilities/test_type_utils.py#L133
But the checking functionality is more strict: https://github.com/ros2/launch/blob/7cc161161eaff98b1f6cd3068148af9ed29fea00/launch/test/launch/utilities/test_type_utils.py#L133

Python behaves in a similar way:
```python3
assert isinstance(3, float) == False
assert float('3') == 3.
```

If we want, we can make the system more flexible in the future.

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+            else:
+                raise
+
+    if bool != type_obj:
```
Done in https://github.com/ros2/launch/pull/438/commits/df07ebd9359bb9ab87ec5893d398b66c6573e085

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+    can_be_str: bool = False,
+) -> StrSomeValueType:
+    """
+    Coerce `value` to `type`.
```
Done, see https://github.com/ros2/launch/pull/438/commits/df07ebd9359bb9ab87ec5893d398b66c6573e085

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+    Check if `value` is of `type`.
+
+    :param value: variable to check.
+    :param data_type: value will be enforced to be an instance of data_type.
```
Yeap, see https://github.com/ros2/launch/pull/438/commits/df07ebd9359bb9ab87ec5893d398b66c6573e085.

ivanpauno commented:
@hidmic I'm still iterating on documentation and tests, but I've at least covered all the functions in `launch.utilities.type_utils`.
Can you take another look?

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+    if types_in_list.issubset({int, float, Substitution}):
+        # list of floats and substitutions
+        if data_type not in (None, float):
+            raise TypeError(err_msg.format(int))
```
done

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,451 @@
+        else:
+            types_in_list.add(None)
+    if None in types_in_list:
+        raise TypeError(
```
I think I've addressed this

ivanpauno commented:
Ok, this one is completely ready now.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+            member_type = str
+        return (
+            all(isinstance(x, valid_types) for x in value[1:]) and
+            member_type in ScalarTypesTuple
```
@ivanpauno meta, safe to ignore: wouldn't doing the check by construction be simpler? That is:

```python
if instance(value, list):
    if not value: 
       return True
    member_type = type(value[0])
    if member_type not in ScalarTypesTuple:
       return False
    valid_types = (member_type, str) if can_be_str else member_type
    return (all(isinstance(x, valid_types) for x in value[1:])
```

ivanpauno replied:
```python3
    member_type = type(value[0])
```
That line is a problem when `can_be_str == True`.
e.g.: `is_instance_of_valid_type(['$(var my_int)', 5], True)` will return `False`. 

The current logic is a bit hard to read, if you have a different way of improving the function let me know.

hidmic replied:
Good point, `member_type = next((type(x) for x in value if type(x) is not str), str)` ?

ivanpauno replied:
Thanks! That looks much nicer.
See https://github.com/ros2/launch/pull/438/commits/ab35da787620e63b4dbe05f59e62b0a08e920bf4.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    Check if `value` is instance of `data_type`.
+
+    :param value: variable to check.
+    :param data_type: class that `value` should be instance of.
```
@ivanpauno nit:
```suggestion
    :param data_type: type that `value` should be an instance of.
```

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    ensure_argument_type(value, str, 'value', 'coerce_to_type')
+    if data_type is None:
+        # No type specified, return best conversion.
+        return convert_as_yaml(value, f"Failed to convert '{value}' using yaml rules")
```
@ivanpauno meta, safe to ignore: looking at this function, it seems unconditionally converting to `yaml` and doing type checks if `data_type` is not `None` may be simpler.

ivanpauno replied:
In some cases, the yaml conversion isn't used.
e.g.: for `int` or `float`

hidmic replied:
Right, but it could be i.e. it'd have the same effect.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    """
+    Normalize a mix of substitution and values.
+
+    This function becomes handy when you need a substitution which result will be coerced to a
```
@ivanpauno typo:
```suggestion
    This function becomes handy when you need a substitution whose result will be coerced to a
```

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    :param value: value to be normalized.
+    :param data_type: `value` can be either an instance of `data_type` or a substitution.
+        In the case of lists, `value` can be either a substitution or a list.
+        In the later case, its item should match the type specified by `data_type` or be a
```
@ivanpauno typo: 
```suggestion
        In the later case, its items should match the type specified by `data_type` or be a
```

ivanpauno replied:
Fixed in https://github.com/ros2/launch/pull/438/commits/cd3fafeca6102d69dc480c3a887ea74a2e659e8a.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    return isinstance(value, ScalarTypesTuple)
+
+
+def is_instance_of(
```
@ivanpauno meta, safe to ignore: this one's name is odd. I'd expect `is_instance_of` to be a variation of `isinstance` that supports a subset of `typing` generics and `is_instance_of_valid_type` to be sugar to call `is_instance_of` with `AllowedTypesTuple`.

ivanpauno replied:
Hmm, I will ignore it for the moment, but we can improve that in the future.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+    """
+    Perform a normalized typed substitution.
+
+    This function becomes handy when you need a substitution which result will be coerced to a
```
@ivanpauno typo:
```suggestion
    This function becomes handy when you need a substitution whose result will be coerced to a
```

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+                perform_substitutions(context, cast(List[Substitution], x)), scalar_type)
+            if is_normalized_substitution(x) else x for x in value
+        ]
+        is_instance_of(output, data_type)
```
@ivanpauno perhaps you wanted:
```suggestion
        assert is_instance_of(output, data_type)
```
instead? There's no side-effect to `is_instance_of` (nor should be IMO).

ivanpauno replied:
https://github.com/ros2/launch/pull/438/commits/cd3fafeca6102d69dc480c3a887ea74a2e659e8a

hidmic commented:
@ivanpauno I was thinking that we should deprecate and not remove what was in `launch.frontend.type_utils` i.e. redirect the call and warn that it has been relocated.

ivanpauno commented on launch/launch/frontend/entity.py:
```diff
@@ -43,46 +45,40 @@ def get_attr(
 
         By default, it will try to return it as an string.
-        `types` states the expected types of the attribute. Type coercion or type checking is
+        `data_type` is the expected types of the attribute. Type coercion or type checking is
```
I would say that the `set of types` part doesn't apply anymore.
I will use the singular:
```suggestion
        `data_type` is the expected type of the attribute. Type coercion or type checking is
```

ivanpauno commented on launch/launch/frontend/entity.py:
```diff
@@ -43,46 +45,40 @@ def get_attr(
-            - a non-uniform list of any scalar type i.e. `list` or `List`;
-            - a `Union` of any of the above;
-            - `List[Entity]`, see below.
+        See :py:obj:`launch.utilities.AllowedTypesTuple` to see what literals `data_type` can be.
```
```suggestion
        See :py:obj:`launch.utilities.AllowedTypesTuple` to see what types are allowed.
```
I will apply the above suggestion.

ivanpauno commented on launch/launch/frontend/parse_substitution.py:
```diff
@@ -112,3 +114,37 @@ def parse_substitution(string_value):
+    When interpolating substitutions in a string, `TextSubstitution` instances are resolved
+    and the original `str` is left.
+
+    :raise: `ValueError` if the result cannot be performed to a valid type.
```
That sounds better.
```suggestion
    :raise: `ValueError` if the result cannot be parsed into a valid type.
```

ivanpauno commented:
> launch.frontend.type_utils

There's still a module called in that way, though I've moved most of the methods.
Also, a lots of the methods have been modified and renamed.

ivanpauno commented on launch/launch/frontend/type_utils.py:
```diff
@@ -12,241 +12,36 @@
+        'bool': bool,
+        'float': float,
+        'int': int,
+        'strs_list': List[str],
```
See https://github.com/ros2/launch/pull/438/commits/0e97e74eba5bb9a945820724830d543c8a0596b8.

ivanpauno commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,565 @@
+        elif len(member_types) > 2:
+            return False
+        else:
+            assert len(member_types) == 1
```
Done https://github.com/ros2/launch/pull/438/commits/c2dacfcdee3fbbed3edd92b84741178b3f1f3e75.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,567 @@
+        hasattr(data_type, '__origin__') and
+        hasattr(data_type, '__args__') and
+        data_type.__origin__ in  # type: ignore
+        (list, List) and  # On Linux/Mac is List, on Windows is list.
```
Note to myself: this is crazy. I wonder why.

ivanpauno replied:
This code was already there before this PR.
I think it wasn't platform dependent, but we where using at that time a different Python version on Windows.

The Python `typing` module is super new and still being developed, so it's not weird that this kind of detail changed.
I guess that the latest version is using `List`, and not `list`.

hidmic replied:
> The Python typing module is super new and still being developed, so it's not weird that this kind of detail changed.

Yeah, that I know. To have a type change on a platform-basis is, nonetheless, curious. Looking at the `typing` module sources, it looks pretty platform-independent to me.

ivanpauno replied:
> Yeah, that I know. To have a type change on a platform-basis is, nonetheless, curious. Looking at the typing module sources, it looks pretty platform-independent to me.

Yes, it's platform independent, but it changed between Python versions.
The comment is misleading, I will correct it.

ivanpauno replied:
Done in https://github.com/ros2/launch/pull/438/commits/9ab3835a7d4f37b212858fd0744ca409bc424cb0

hidmic replied:
> Yes, it's platform independent, but it changed between Python versions.

Aha! Now it makes sense.

hidmic commented on launch/launch/utilities/type_utils.py:
```diff
@@ -0,0 +1,567 @@
+    To avoid confusions, the passed value will always be interpreted as a single substitution
+    if possible, like in the above example (which will resolved to `'asdbsd'`).
+    To make it a list of substitutions:
+    `['asd', [TextSubstitution(text='bsd')]]`
```
+1

hidmic commented on launch/test/launch/utilities/test_type_utils.py:
```diff
@@ -0,0 +1,578 @@
+    assert not is_instance_of_valid_type_impl(test_is_instance_of_valid_type)
+    assert not is_instance_of_valid_type_impl({'key': 'value'})
+
+    assert is_instance_of_valid_type_impl(1, True)
```
@ivanpauno nit: here and below, using a `can_be_str` kwarg helps readability.

ivanpauno replied:
done.
See https://github.com/ros2/launch/pull/438/commits/150deca7d9c9b37f030cce350450dadd0c623b0f and https://github.com/ros2/launch/pull/438/commits/99cc84c1e1d89ad706f2cbc2ddd3d1acc021ab98.

hidmic commented on launch/test/launch/utilities/test_type_utils.py:
```diff
@@ -0,0 +1,578 @@
+    assert coerce_to_type_impl('1', data_type=int) == 1
+    assert coerce_to_type_impl('1000', data_type=int) == 1000
+
+    assert coerce_to_type_impl('1', data_type=float) == 1.0
```
@ivanpauno hmm, so coercion is flexible enough to accept an integer representation when `float` is explicitly given as the expected data type, but otherwise fails (even when a `List[float]` is specified)? I find it surprising at least.

ivanpauno replied:
> @ivanpauno hmm, so coercion is flexible enough to accept an integer representation when float is explicitly given as the expected data type, but otherwise fails (even when a List[float] is specified)? I find it surprising at least.


I think that when `List[float]` is explicitly specified we could accept both ints and floats members.
IMO, we can make this more permissive in the future.

hidmic replied:
Yeap. I'd be slightly inclined towards making it consistent (uniform) no, but I won't block on it.

ivanpauno commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11679)](http://ci.ros2.org/job/ci_linux/11679/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6792)](http://ci.ros2.org/job/ci_linux-aarch64/6792/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9543)](http://ci.ros2.org/job/ci_osx/9543/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=11644)](http://ci.ros2.org/job/ci_windows/11644/)

dirk-thomas commented:
It seems that this change resulted in a regression across all platforms: see e.g. https://ci.ros2.org/view/nightly/job/nightly_linux_debug/1641/testReport/test_launch_ros.test.test_launch_ros.frontend/test_node_frontend/test_launch_remapping_xml/

@ivanpauno Please address asap.
